veracruz-mcu-client/vc.c

```diff
@@ -0,0 +1,577 @@
+    if (res < 0) {
+        printf("base64_encode failed (%d)\n", res);
+        return res;
+    }
+    data_len += res;
+
+    // send data over HTTP POST
+    #ifdef VC_DUMP_INFO
+        printf("sending to %s:%d:\n",
+                VC_SERVER_HOST,
+                VC_SERVER_PORT);
+        xxd(vc->send_buf, data_len);
+    #endif
+    printf("veracruz_server -> POST /runtime_manager, %d bytes\n", data_len);
+    ssize_t recv_len = http_post(
+            VC_SERVER_HOST,
+            VC_SERVER_PORT,
+            "/runtime_manager",
+            vc->send_buf,
+            data_len,
+            vc->recv_buf,
+            VC_RECV_BUFFER_SIZE);
+    if (recv_len < 0) {
+        printf("http_post failed (%d)\n", recv_len);
+        return recv_len;
+    }
+    printf("veracruz_server <- 200 OK, %d bytes\n", recv_len);
+
+    if (recv_len == 0) {
+        // done, recieved nothing
+        return len;
+    }
+
+    #ifdef VC_DUMP_INFO
+        printf("ssl session: recv:\n");
+        xxd(vc->recv_buf, recv_len);
+    #endif
+
+    // null terminate to make parsing a bit easier
+    vc->recv_buf[recv_len] = '\0';
+
+    // we have a bit of parsing to do, first decode session id
+    const uint8_t *parsing = vc->recv_buf;
+    bool session_id_was_zero = vc->session_id == 0;
+    vc->session_id = strtol(parsing, (char **)&parsing, 10);
+    if (parsing == vc->recv_buf) {
+        printf("failed to parse session id\n");
+        return -EILSEQ;
+    }
+    // skip space
+    parsing += 1;
+    if (session_id_was_zero) {
+        printf("\033[32mestablished session id:\033[m %d\n", vc->session_id);
+    }
+
+    // parse out base64 blobs, shuffling to front of our buffer
+    uint8_t *parsed = vc->recv_buf;
+    while (parsing < &vc->recv_buf[recv_len]) {
+        int i = 0;
+        while (parsing[i] && parsing[i] != ' ') {
+            i += 1;
+        }
+
+        res = base64_decode(
+                parsing, i,
+                parsed, &vc->recv_buf[recv_len]-parsed);
+        if (res < 0) {
+            printf("base64_decode failed (%d)\n", res);
+            return res;
+        }
+
+        parsing += i;
+        parsed += res;
+
+        // skip space
+        if (parsing[0] == ' ') {
+            parsing += 1;
+        }
+    }
+
+    vc->recv_pos = vc->recv_buf;
+    vc->recv_len = parsed - vc->recv_buf;
+
+    #ifdef VC_DUMP_INFO
+        printf("ssl session: parsed:\n");
+        xxd(vc->recv_pos, vc->recv_len);
+    #endif
+
+    // done!
+    return len;
+}
+
+static ssize_t vc_rawrecv(void *p,
+        uint8_t *buf, size_t len,
+        uint32_t timeout) {
+    vc_t *vc = p;
+
+    if (vc->recv_len == 0) {
+        // no data available? since we communicate over POSTs,
+        // we'll never have data available
+        printf("recv timeout\n");
+        return MBEDTLS_ERR_SSL_TIMEOUT;
+    }
+
+    size_t diff = (len < vc->recv_len) ? len : vc->recv_len;
+    memcpy(buf, vc->recv_pos, diff);
```

Comment:
What happens when `vc->recv_len > len`? Do we just return the truncated data?

Reply:
Currently yes, can change this to an error

Reply:
Wait no, sorry, this should be correct.

It's providing a stream/read/recv API, so if the buffer (`len`) is smaller than the available data `vc->recv_len`, then we return the truncated data but keep the rest of the data in our internal buffer. mbedtls or whatever consumer is on top should call vc_rawrecv repeatedly until all data is consumed.